### PR TITLE
[peer storage] post ready

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**

### CODE REVIEW

File: tikv/raftstore/keys.go [hunk 1]
```diff
@@ -122,6 +123,22 @@ func OriginKey(key []byte) []byte {
 	return key[1:]
 }
 
+// Get the `start_key` of current region in encoded form.
+func EncStartKey(region *metapb.Region) []byte {
+	// only initialized region's start_key can be encoded, otherwise there must be bugs
+	// somewhere.
+	y.AssertTruef(len(region.Peers) > 0, "region is not initialized")
```
Comment (coocood) [inline tikv/raftstore/keys.go]: I prefer `y.Assert` than `y.AssertTruef` for better performance.

File: tikv/raftstore/peer_storage.go [hunk 2]
```diff
@@ -19,15 +19,32 @@ import (
...
+	SnapState_Relax SnapStateType = 0 + iota
 	SnapState_Generating
 	SnapState_Applying
 	SnapState_ApplyAborted
 )
 
+type SnapState struct {
+	StateType SnapStateType
+	Status    *JobStatus
+	Receiver  chan<- eraftpb.Snapshot
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: better use pointer type in chan.

File: tikv/raftstore/peer_storage.go [hunk 3]
```diff
@@ -134,6 +151,32 @@ func (ec *EntryCache) compactTo(idx uint64) {
...
+	///
+	/// The deletion may and may not succeed.
+	RegionTaskType_Destroy
+)
+
+type RegionTask struct {
+	RegionId uint64
+	TaskType RegionTaskType
+	Notifier chan<- eraftpb.Snapshot
+	Status   *JobStatus
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: We don't need to use pointer.

Comment (zhangjinpeng1987) [reply tikv/raftstore/peer_storage.go]: In TiKV, it is a atomic usize shared by multiple threads.

Comment (coocood) [reply tikv/raftstore/peer_storage.go]: I know, we can use int64 type to do atomic operations.

File: tikv/raftstore/peer_storage.go [hunk 4]
```diff
@@ -134,6 +151,32 @@ func (ec *EntryCache) compactTo(idx uint64) {
...
+	/// Destroy data between [start_key, end_key).
+	///
+	/// The deletion may and may not succeed.
+	RegionTaskType_Destroy
+)
+
+type RegionTask struct {
+	RegionId uint64
+	TaskType RegionTaskType
+	Notifier chan<- eraftpb.Snapshot
```
Comment (coocood) [inline tikv/raftstore/peer_storage.go]: pointer

File: tikv/raftstore/peer_storage.go (hunk 3 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/peer_storage.go]: Yes, JobStatus is already int/int64.

Comment (zhangjinpeng1987) [reply tikv/raftstore/peer_storage.go]: And `JobStatus` shared by `RegionTask` and `SnapState` which in different threads.

**Comment by coocood** [comment]
LGTM

**Pull request merged**